measurements/pages/__init__.py

```diff
@@ -204,34 +208,69 @@ def files_in_country(country_code):
+        current_app.logger.debug("  frames_off: %d" % frames_off)
+        current_app.logger.debug("  frames_size: %d" % frames_size)
+        current_app.logger.debug("  intra_off: %d" % intra_off)
+        current_app.logger.debug("  intra_size: %d" % intra_size)
+        try:
+            r = requests.get(url, headers={"Range": "bytes={}-{}".format(frames_off, frames_off + frames_size)},
+                                stream=True)
+            streamed_data = 0
+            for chunk in lz4framed.Decompressor(r.raw):
+                current_app.logger.debug("chunking away 1")
+                d = chunk
+                current_app.logger.debug("chunking away 2")
+                if streamed_data == 0 and intra_off > 0:
+                    d = chunk[intra_off:]
+                if streamed_data > (intra_size + len(d)):
+                    d = chunk[:(streamed_data - intra_size)]
+                yield d
+                streamed_data += len(d)
+            current_app.logger.debug("sent: %d" % streamed_data)
+        except Exception as exc:
+            current_app.logger.error("failed to fetch streamed data: %s" % exc)
+            raise HTTPException("Failed to fetch data")
+    return generator
+
+@pages_blueprint.route('/files/download/<path:textname>')
+def files_download(textname):
+    q = current_app.db_session.query(
+            Measurement.id.label('m_id'),
+            Measurement.report_no.label('report_no'),
+            Measurement.frame_off.label('frame_off'),
+            Measurement.frame_size.label('frame_size'),
+            Measurement.intra_off.label('intra_off'),
+            Measurement.intra_size.label('intra_size'),
+            Report.report_no.label('r_report_no'),
+            Report.autoclaved_no.label('r_autoclaved_no'),
+            Autoclaved.filename.label('a_filename'),
+            Autoclaved.autoclaved_no.label('a_autoclaved_no'),
+    ).filter(Report.textname == textname) \
+        .join(Report, Report.report_no == Measurement.report_no) \
+        .join(Autoclaved, Autoclaved.autoclaved_no == Report.autoclaved_no) \
+        .order_by(Measurement.frame_off.asc(), Measurement.intra_off.asc())
+
+    msmts = q.all()
+    if len(msmts) == 0:
+        current_app.logger.debug("Could not find %s" % textname)
         raise NotFound("No file with that filename found")
 
-    bucket_date, filename = report.textname.split("/")
-    return redirect(
-        get_download_url(
-            current_app,
-            bucket_date,
-            filename
-        )
-    )
+    autoclaved_filename = msmts[0].a_filename
+    intra_off = msmts[0].intra_off
+    intra_size = 0
+    for msmt in msmts:
+        if autoclaved_filename != msmt.a_filename:
+            raise HTTPException("Autoclaved file is spanned across multiple files")
+        intra_size += msmt.intra_size
+    frames_off = msmts[0].frame_off
+    frames_size = msmts[-1].frame_off - frames_off + msmts[-1].frame_size
```

Comment:
This means that measurements:

- do not span autoclaved files
- start at `frames_off`
- I do not fully understand `frame_size` (is the length starting from off, like in `struct iovec`)?
- inside a frame there is an offset `intra_off`
- I don't get `intra_size` (shouldn't it be the size of a frame minus the offset?)

Reply:
Ah, maybe it's like this:

`frame_off < intra_off < frame_off + intra_size < frame_off + frame_size`?

Reply:
I guess this also assumes that frames returned by the database are consecutive. Perhaps this should be documented in a comment, to help the reader understand?

(I guess I am having troubles because I'm not very familiar with the way the new pipeline stores data.)

Reply:
`frames_size` and `intra_size` are just utility variables I create to make the job easier of figuring out what I need.

`frame_size`: (notice how it has an `s` at the end versus `frame_off` having no `s`) is the total size of all the frames. If the report file I need to fetch spans multiple frames I will consider `frames_size` the total of all consecutive frames (we have the guarantee from the data format that measurements in the same report end up in consecutive frames)

`intra_size` is similar to `frames_size` except it's how much uncompressed data I need to fill up the report with goodness.

Reply:
Mumble: so, it's correct to say that `frames` variables are compressed and `intra` variables are uncompressed? If so, would you mind renaming them? I am a bit lost with the current naming.

Reply:
measurements & reports, do not span autoclaved files
LZ4 frame containing measurement starts at frames_off
Every LZ4 frame is something like ~256 Kbytes
measurements always fits single LZ4 frame
single LZ4 frame may have several measurements and other stuff (e.g. tar header)
reports may be split over several LZ4 frames

Reply:
Thanks, @darkk, I understand. So, basically, we have two base + offset couples. One for compressed and one for uncompressed data. Correct?

Reply:
(I guess `base + offset` is my mental bias, because it reminds me of `iovec`)

Reply:
> `frames_size` ... `frame_size`

Just a generic request. Please, avoid variable names having different meaning and Levenshtein distance of ~1. I'm not that smart and careful to see every single letter :)

Reply:
> So, basically, we have two base + offset couples. One for compressed and one for uncompressed data. Correct?

Yep. These offsets are per-measurement to be able to fetch individual measurement from compressed file and report is always stored as consequent stream of bytes, so only the first and the last quadruple matter.

Reply:
> Mumble: so, it's correct to say that frames variables are compressed and intra variables are uncompressed? If so, would you mind renaming them? I am a bit lost with the current naming.

I used this naming as it's consistent with the naming inside of the DB schema. I think naming them differently would not lead to further clarity as it's something fairly specific (i.e. you need to know the schema of the DB to know what this is doing).

I can renames `frames_size` to `all_frames_size` (and `intras_size` to `all_intras_size`) if that makes it clearer though.

Reply:
> I used this naming as it's consistent with the naming inside of the DB schema.

Okay, this is a good argument.

> I can renames frames_size to all_frames_size (and intras_size to all_intras_size) if that makes it clearer though.

Keeping same names as database is probably wiser.

I think that a comment just saying that there is a `offset` and a `size` for compressed data and similarly a `offset` and a `size` for uncompressed data will be enough to clarify.